lib/hotel.rb

```diff
@@ -0,0 +1,184 @@
+
+	  def initialize
+		#this array will hold room objects for each room in the hotel
+		@rooms = []
+		#this is a hash that maps room numbers to reservations for that room
+		@reservations_by_room = {}
+		#this is a hash that maps dates to reservations on that date
+		@reservations_by_date = {}
+		#this array holds all of the blocks
+		@room_blocks = []
+		#this array holds all of the rooms that are currently a part of some block
+		@rooms_currently_in_block = []
+		20.times do |i|
+			@rooms << Hotels::Room.new(i)
+			@reservations_by_room[i] = []
+		end
+	  end
+	  
+	  def reserve_room(room_number, date)
+		#this will check if the given room number is a memeber of any block
+		if @rooms_currently_in_block.include?(room_number)
+			#iterate through each block
+			@room_blocks.each do |current_block|
+				#iterate through this blocks room variable
+				current_block.rooms.each do |current_room_in_block|
+					#if this room number matches the room number in question
+					if current_room_in_block.room_number == room_number
+						#checks if the start dates and end dates conflict,
+						#raises errors accordingly
+						if date.start_date > current_block.start_date && 
+						date.start_date < current_block.end_date
+							raise ArgumentException, "This room is part of a block. Change your start date"
+						elsif date.end_date > current_block.start_date
+							raise ArgumentException, "This room is part of a block. Change your end date"
+						end
+					end
+				end
+			end
+		end
+		
+		#pulls all array of all reservations of a given room
+		current_reservations = @reservations_by_room[room_number]
+		current_reservations.each do |i|
+			#if the requested start date is before the current reservations start date
+			if date.start_date < i.start_date  
+				#checks to see if the end date is also before the start date
+				if (date.end_date - i.start_date).to_i > 0 
+					raise ArgumentError, "This room is booked during the dates requested"
+				end
+				#if the start dates are the same, or the start date is less than the end date
+			elsif date.start_date == i.start_date || (date.start_date - i.end_date).to_i < 0 
+				raise ArgumentError, "This room is booked during the dates requested"
+			end
+		end
+		
+		#makes new reservation
+		new_reservation = Hotels::Reservation.new(room_number, date)
+		@reservations_by_room[room_number] << new_reservation
+		
+		#if this day already has a reservation
+		if @reservations_by_date.key?(date.start_date)
+			#add it to the array
+			@reservations_by_date[date.start_date] << new_reservation
+		else
+			#other wise make a new array and add it
+			@reservations_by_date[date.start_date] = [new_reservation]
+		end
+			
+	  end
+	  
+	  
+	  def get_all_rooms
+		return @rooms
+	  end
+	  
+	  def get_open_rooms(date)
+		open_rooms = []
+		#makes a boolean array flagging all rooms as available
+		20.times do |i|
+			open_rooms[i] = true
```

Comment:
There are a few places in the codebase where we are using this technique of creating an array of booleans that maps to an array of other values, then we loop through the values and set the booleans according to some conditional checks, and finally we create a new array from all of the values which are now mapped to a true boolean.

This is a pretty common pattern in programming, generally referred to _filtering_, and Ruby's Enumerable class has a method for it: [`Enumerable#select`](https://ruby-doc.org/core-2.2.3/Enumerable.html#method-i-select).

We could reduce the code needed for this method by using it thusly:
```ruby
def get_open_rooms(date)
  available_rooms = reservations_by_date.values.flatten.select do |res|
    # This should be a valid check for overlapping date ranges
    (res.start_date <= date.start_date && res.end_date > date.start_date) ||
    (date.start_date <= res.start_date && date.end_date > res.start_date)
  end.map(&:room_number)

  return available_rooms
end
```

As per my other comments, the above implementation could be simplified further by using a single `@reservations` array, and by implementing the date check as something like `DateRange#overlap?`.